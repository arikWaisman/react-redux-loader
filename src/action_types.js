export const INTERFACE_REQUEST = 'react-redux-modal/INTERFACE_REQUEST';
export const INTERFACE_COMPLETE = 'react-redux-modal/INTERFACE_COMPLETE';
export const INTERFACE_ERROR = 'react-redux-modal/INTERFACE_ERROR';

export const SHOW_LOADER = 'react-redux-modal/SHOW_LOADER';
export const HIDE_LOADER = 'react-redux-modal/HIDE_LOADER';

export const SHOW_INLINE_LOADER = 'react-redux-modal/SHOW_INLINE_LOADER';
export const HIDE_INLINE_LOADER = 'react-redux-modal/HIDE_INLINE_LOADER';
