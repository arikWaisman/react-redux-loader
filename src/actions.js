import {
  SHOW_LOADER,
  HIDE_LOADER,
  SHOW_INLINE_LOADER,
  HIDE_INLINE_LOADER,
  INTERFACE_REQUEST,
  INTERFACE_COMPLETE,
} from './action_types';

export const showSyncLoader = (args = {}) => {
  const { type = null, shouldTimeOut = null, timeOut = null } = args;
  return {
    type: SHOW_LOADER,
    payload: {
      type,
      ...(shouldTimeOut ? { shouldWaitToDisplay: true, timeOut } : {}),
    },
  };
};

export const hideSyncLoader = () => ({
  type: HIDE_LOADER,
});

export const showInlineLoader = (key = null) => ({
  type: SHOW_INLINE_LOADER,
  payload: {
    key,
  },
});

export const hideInlineLoader = (key = null) => ({
  type: HIDE_INLINE_LOADER,
  payload: {
    key,
  },
});

export const interfaceRequest = (args = {}) => {
  const { type = null, shouldTimeOut = null, timeOut = null } = args;
  return {
    type: INTERFACE_REQUEST,
    payload: {
      type,
      ...(shouldTimeOut ? { shouldWaitToDisplay: true, timeOut } : {}),
      async: true,
    },
  };
};

export const interfaceComplete = () => ({
  type: INTERFACE_COMPLETE,
});
