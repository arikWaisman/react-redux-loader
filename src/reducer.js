import {
  INTERFACE_REQUEST,
  INTERFACE_COMPLETE,
  INTERFACE_ERROR,
  SHOW_LOADER,
  HIDE_LOADER,
  SHOW_INLINE_LOADER,
  HIDE_INLINE_LOADER,
} from './action_types';

const INITIAL_STATE = {
  requests: 0,
  showLoader: null,
  error: null,
  loaderType: null,
  shouldWaitToDisplay: false,
  timeOut: null,
  async: false,
  inline: [],
};

// keep loader type if already one open
const getLoaderType = (requestedType = null, state) =>
  state.requests > 0 ? state.loaderType : requestedType;

// remove an inline loader from inline list
const updateInlineLoaders = (loaderToRemove, state) => {
  const indexToRemove = state.inline.indexOf(loaderToRemove);

  return state.inline.filter((key, index) => index !== indexToRemove);
};

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case INTERFACE_REQUEST:
      return {
        ...state,
        requests: state.requests + 1,
        loaderType: getLoaderType(action.payload.type, state),
        shouldWaitToDisplay:
          action.payload.shouldWaitToDisplay || state.shouldWaitToDisplay,
        timeOut: action.payload.timeOut || state.timeOut,
        async: true,
      };

    case INTERFACE_COMPLETE:
      return {
        ...state,
        requests: state.requests - 1,
        ...(state.requests - 1 > 0
          ? {}
          : {
              shouldWaitToDisplay: false,
              timeOut: null,
              async: false,
              showLoader: false,
            }),
      };

    case SHOW_LOADER:
      return {
        ...state,
        showLoader: true,
        loaderType: action.payload.type,
        shouldWaitToDisplay:
          action.payload.shouldWaitToDisplay || state.shouldWaitToDisplay,
        timeOut: action.payload.timeOut || state.timeOut,
      };

    case HIDE_LOADER:
      return { ...state, showLoader: false, async: false, loaderType: null };

    case SHOW_INLINE_LOADER:
      return { ...state, inline: [...state.inline, action.payload.key] };

    case HIDE_INLINE_LOADER:
      return {
        ...state,
        inline: updateInlineLoaders(action.payload.key, state),
      };

    case INTERFACE_ERROR:
      return { ...state, error: action.error };

    default:
      return state;
  }
}
