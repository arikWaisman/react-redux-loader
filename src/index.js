import LoaderContext from './loader_context';
import * as actions from './actions';
import * as actionTypes from './action_types';
import loaderReducer from './reducer';
import makeLoaderRoot from './components/loader_root';
import LoaderFactory from './components/loader_factory';

export {
  actions,
  actionTypes,
  makeLoaderRoot,
  loaderReducer,
  LoaderFactory,
  LoaderContext,
};
