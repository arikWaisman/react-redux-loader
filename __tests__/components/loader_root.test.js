import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { render, cleanup } from 'react-testing-library';

import { makeLoaderRoot } from '../../src';

const mockStore = configureStore([]);

describe('Loader root', () => {
  beforeEach(cleanup);

  test('renders and mount html', () => {
    const store = mockStore({ loader: {} });
    const LoaderRoot = makeLoaderRoot(store);

    const fakeMountToDom = jest.spyOn(LoaderRoot.prototype, 'mountToDOM');
    const wrapper = shallow(<LoaderRoot />);
    const { getByTestId } = render(<LoaderRoot />);

    // make sure our initial render method returns correctly
    expect(toJson(wrapper)).toMatchSnapshot();

    // make sure our mounting function is called
    expect(fakeMountToDom).toHaveBeenCalled();

    // make sure our mounted html returns correctly
    expect(getByTestId('loader-root')).toMatchSnapshot();
  });
});
