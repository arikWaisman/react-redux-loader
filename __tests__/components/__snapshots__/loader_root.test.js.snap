// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader root renders and mount html 1`] = `<noscript />`;

exports[`Loader root renders and mount html 2`] = `
<div
  class="full-loader"
  data-testid="loader-root"
/>
`;
